source/unified-test-format/unified-test-format.rst

```diff
@@ -1798,6 +1897,119 @@ Use a ``listIndexes`` command to check whether the index exists. Note that it is
+  - ``time``: the number of (floating-point) seconds since the Unix epoch
+    when the error was encountered.
+
+- ``storeFailuresAsEntity``: if specified, the runner MUST handle failures
+  arising during sub-operation execution and append a document with failure
+  information to the array stored in the specified entity. If
+  not specified, the runner MUST treat failures as errors, and either
+  handle them following the logic described in ``storeErrorsAsEntity``
+  or cause them to terminate execution, if ``storeErrorsAsEntity`` is not
+  specified. The failure document MUST contain the following fields:
+  
+  - ``error``: the textual description of the failure encountered.
+  - ``time``: the number of (floating-point) seconds since the Unix epoch
+    when the failure was encountered.
+
+- ``storeIterationsAsEntity``: if specfied, the runner MUST keep track of
+  the number of iterations of the loop performed, and store that number
+  in the specified entity.
+
+A *failure* is when an operation performed by the test runner has a
+different result from what the test expected. For example, a test
+containing an expectation that a particular document has a particular value
+in a particular field, and the actual value being different, would
+be a failure. An *error* is any other type of error. For example,
+attempting to invoke an undefined operation would be an error.
+
+This specification permits the test runner to report some failures as errors
+and some errors as failures. When the test runner stores errors and
+failures as entities it MAY classify conditions as errors and failures in
+the same way as it would when used in the driver's test suite.
+This includes reporting all errors as failures or all failures as errors.
+
+If the test runner does not distinguish errors and failures in its reporting,
+it MAY report both conditions under either category, but it MUST report
+any given condition in at most one category.
+
+The same entity name MAY be used for both ``storeErrorsAsEntity``
+and ``storeFailuresAsEntity`` options. The test runner MUST support
+writing both errors and failures to the same entity. If the same entity is
+specified, the errors and failures MUST be written in chronological order
+to the extent possible, interleaving the error and failure entries
+if necessary. Note that the textual description field for both errors
+and failures is ``error``.
+
+The following termination behavior MUST be implemented by the unified test
+runner:
+
+- There MUST be a way to request termination of the loops. This request
+  will be made by the Atlas testing workload executor in response to
+  receiving the termination signal from Astrolabe.
+  
+- When the termination request is received, the workload executor MUST
+  stop looping. The current loop iteration SHOULD complete to its natural
+  conclusion (success or failure).
+
+- After the termination request is received, the runner MUST NOT start
+  any further loops.
```

Comment:
Can we replace this with two requirements for test file structure?

 * The list of operations for a test MUST NOT include multiple `loop` operations. Furthermore, `loop` operations MUST NOT be nested within other `loop` operations either directly or indirectly (e.g. `loop` within `withTransaction` within `loop`).

I don't expect to require the test runner to enforce this (folks could if they so choose) but it would provide guidance for how we write tests. We could always revise this later if a use case arises, but for now I think this will free us from having to worry about some edge cases.

Reply:
Let's leave this as-is and just add "any further loops in either the current test or subsequent tests for the lifetime of the test runner".

Reply:
I updated the termination behavior to match our discussion earlier. The case of receiving termination request prior to looping is explicitly handled and the success upon receiving termination request is no longer mandated.